### ISSUE TITLE
Added microsecond to formatTime

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3513]
|<img alt="wangchun's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wangchun reported|
|-|-|
|Trac ID|trac#3513|
|Type|enhancement|
|Created|2008-10-29 12:39:18Z|
|Branch|https://github.com/twisted/twisted/tree/datetime-strftime-3513|

The current formatTime implementation cannot format some time zone string correctly.

For busy servers, timestamp with microsecond is useful to administrators understand what was happened from log files.

Currently, user cannot give format string to reproduce the default format. We should make the time format more flexible and customizable by the user. Any ideas?

Attachments:

* [log.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5c1/5c12d9c16c870ca2dfa27e8c8a62b734b7a07cbc/a8bf44c1276b2ac10ec0acf104bfd8996b0a5763.diff) (3135 bytes) - added by wangchun on 2008-10-29 12:39:31Z - 
* [log-tzinfo-3513.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5c1/5c12d9c16c870ca2dfa27e8c8a62b734b7a07cbc/82942a8ff036044325a5d7beacc17df48105aaf6.diff) (14452 bytes) - added by zseil on 2008-11-17 03:33:31Z - add timezone info support to LogFileObservers

<details><summary>Searchable metadata</summary>

```
trac-id__3513 3513
type__enhancement enhancement
reporter__wangchun wangchun
priority__normal normal
milestone__ 
branch__branches_datetime_strftime_3513 branches/datetime-strftime-3513
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__twistd twistd
time__1225283958000000 1225283958000000
changetime__1356913248000000 1356913248000000
version__None None
owner__exarkun exarkun
cc__exarkun cc__zseil
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

You can subclass `FileLogObserver` and override `formatTime`.  That wouldn't be much harder than replacing the time format with a new string.

Most cases won't benefit from having microseconds in the log, so I don't think that data should be included by default.

Maybe `time.strftime` should gain a new format character for this value?

**Comment by twisted-trac** [comment]
|<img alt="wangchun's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|wangchun commented|
|-|-|

You are right, but the code would look a bit hacking, since FileLogObserver is hard coded in startLogging.

Perhaps we can find some way to do it without having to subclass the observer. There is a Formatter class in Python's logging module.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

[#11359](https://github.com/twisted/twisted/issues/11359) is a ticket for making it possible to customize the logging configuration used by `twistd`.  That would address the problem of `FileLogObserver` being hard coded in `startLogging`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I filed [#3515](https://github.com/twisted/twisted/issues/3515) for the small negative tzoffset bug.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_5"></a>|@zseil commented|
|-|-|

`datetime.datetime.strftime()` in Python 2.6 accepts a new `%f` format character, which expands to the number of microseconds. So I used this as an excuse to convert all time manipulation in `FileLogObserver` to `datetime` module objects. The patch adds a new method `getTimezoneInfo` to `FileLogObserver`, which controls the current local time. It also always uses the `datetime.strftime()` method for time formating. The default time format is unchanged, but now that [#3534](https://github.com/twisted/twisted/issues/3534) has landed it should be easier to change `timeFormat` used by `twistd`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

The patch attached as little to do with the current ticket. It would be better to open a new one.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_8"></a>|@zseil commented|
|-|-|

Ok, I opened [#3546](https://github.com/twisted/twisted/issues/3546).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

[#6219](https://github.com/twisted/twisted/issues/6219) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [36644]) Branching to 'datetime-strftime-3513'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Just switched to using `datetime.datetime.strftime` instead of `time.strftime`.  The former seems to be a superset of the latter, just adding `%f` (microseconds) and `%z` (timezone offset).  It's probably a bit slower too.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/datetime-strftime-3513)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince set owner to @exarkun|
|-|-|

Apparently ```timedelta.total_seconds``` is only available in python 2.7:
```
exceptions.AttributeError: 'datetime.timedelta' object has no attribute 'total_seconds'
```
so that hunk should be reverted.

Also, the documentation of ```formatTime``` needs to updated to reflect the changed underlying function.

Please commit after those changes and a green buildbot run.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [36699]) Merge datetime-strftime-3513

Author: exarkun
Reviewer: tom.prince
Fixes: [#3513](https://github.com/twisted/twisted/issues/3513)

Switch from `time.strftime` to `datetime.datetime.strftime` for formatting
log message timestamps, adding support for microseconds and timezone offsets.
